// dvd-service-js-ajax-file-upload/oss.js
import $ from 'jquery';
import OSS from 'ali-oss';
import uuid from 'uuid/v1';
import Img from 'dvd-base-js-img';
import date from 'dvd-base-js-date';
import type from 'dvd-base-js-type';
import popup from 'dvd-service-js-popup';
import encrypt from 'dvd-service-js-encrypt';

export default {
  /* 接口获取临时账户信息 */
  getSts() {
    return new Promise((resolve, reject) => {
      $.ajax({
        cache: false,
        async: true,
        url: '/api/mg/images/osssts/getToken?_=' + Date.now(),
        type: 'post',
        dataType: 'json',
        data: encrypt.ajax({}),
        success(response) {
          try {
            if (response.code === 0) {
              resolve(response.data);
            } else {
              reject(response && response.data && response.data.msg);
            }
          } catch (err) {
            reject(response && response.data && response.data.msg);
          }
        },
        error(error) {
          reject(error);
        },
      });
    });
  },

  /* 初始化oss client */
  initClient() {
    return new Promise((resolve, reject) => {
      this.getSts().then((sts) => {
        // 初始化oss client
        resolve(new OSS.Wrapper({
          region: 'oss-cn-beijing',

          // bucket: 'mamaj-beta-oss',  // 测试bucket

          bucket: 'mamaj-oss',    // 正式bucket

          // 凯强新建子账号
          // accessKeyId: 'LTAITipazbYIRzeA',
          // accessKeySecret: '3tsSyCUZkvenolzeBMI2QKu7Gwgibb',

          // STS账号
          accessKeyId: sts.keyId,
          accessKeySecret: sts.keySecret,
          stsToken: sts.token,
        }));
      }, (err) => {
        reject(err);
      })
    });
  },

  /* 根据文件name字段生成随机文件名 */
  generateFileName(file, type) {
    // 取type优先级：type > file.type >  file.name > ''
    type = type || Img.getImgExt(file) || (function () {
      if(file.name) {
        // 长度大于1代表有有扩展名
        let arr = file.name.split('.');
        if (arr.length > 1) {
          return arr[arr.length - 1];
        }
        return null;
      }
    })() || '';

    return `${date.format(Date.now(), 'yyyyMMdd-hhmmss-SSS')}-${uuid()}.${type}`;
  },

  /* 上传文件到OSS */
  upload (dirName = 'test' /* 文件夹名 */, file /* 文件对象 */) {
    // 参数监测
    if (dirName == 'test') {
      console.warn(`您正在向${dirName}目录上传文件，test目录随时有可能被删除，请您及时修改目录名，确保在上线后不再向test目录上传文件`);
    }

    let fileName = this.generateFileName(file);
    return new Promise((resolve, reject) => {
      this.initClient().then((client) => {
        // 开始分片上传
        client.multipartUpload(`/fe/upload/${dirName}/${fileName}`, file).then((data) => {
          let url = data.res.requestUrls[0];
          url = url.replace('mamaj-oss.oss-cn-beijing.aliyuncs.com', 'pic.davdian.com');
          url = url.split('?uploadId=')[0];
          resolve(url);
        }).catch((err) => {
          reject(err);
        });
      }, (err) => {
        reject(err);
      });
    });
  }

}
