// dvd-service-js-native/dist/dvd-service-js-native.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _LiveVideo;

var _dvdBaseJsUa = require('dvd-base-js-ua');

var _dvdBaseJsUa2 = _interopRequireDefault(_dvdBaseJsUa);

var _dvdServiceJsPopup = require('dvd-service-js-popup');

var _dvdServiceJsPopup2 = _interopRequireDefault(_dvdServiceJsPopup);

var _dvdBaseJsRuntime = require('dvd-base-js-runtime');

var _dvdBaseJsRuntime2 = _interopRequireDefault(_dvdBaseJsRuntime);

var _jquery = require('jquery');

var _jquery2 = _interopRequireDefault(_jquery);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; } /**
                                                                                                                                                                                                                   * @module dvd-service-js-native
                                                                                                                                                                                                                   * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-native.git)
                                                                                                                                                                                                                   * @see http://wiki.bravetime.net/pages/viewpage.action?pageId=8192341
                                                                                                                                                                                                                   */


var config = {
  // 默认分享信息
  defaultShareInfo: {
    title: 'MAMA+|大V店', // 分享标题
    desc: 'MAMA+|大V店', // 分享描述
    link: _dvdBaseJsRuntime2.default.isClient() ? window.location.href : 'https://bravetime.davdian.com', // 分享链接
    imgUrl: 'http://pic.davdian.com/free/index0925_icon1.png?x-oss-process=image/resize,m_fill,w_80' // 分享图标
  }
};

// 默认标题栏
var defaultInitHead = {
  showHead: 1, // 是否展示头部
  backOnHead: 1, // 头部返回按钮
  homeOnHead: 0, // 头部首页按钮
  shareOnHead: 0, // 头部分享按钮
  btnOnHead: 0, // 头部文字按钮
  btnText: '', // 头部文字按钮文字
  btnLink: '', // 头部文字按钮链接
  showFoot: 0 // 是否展示底部
};

// 默认标题栏
var defaultSetHead = {
  title: '',
  backBtn: '1', // 0表示头部不展示返回按钮，1表示展示
  homeBtn: '0', // 0表示头部不展示首页按钮，1表示展示
  shareBtn: '0' //0表示头部不展示分享按钮，1表示展示
  // shareMoney: '',    //0表示分享佣金为0不显示分享赚钱按钮，非0：展示分享赚钱按钮，在3.9.1之前用该字段
  // shareMoneyStr: '',   //在3.9.1中会用该字段
  // rightBtn: {    // rightBtn会覆盖其他字段
  //   text: '',
  //   textColor: '#ff4a7d',
  //   action: ''
  // },
};

/**
 * 功能: 拼接调用native的通信协议(核心方法)
 * 用法:
 * call({
      v: '3.1.0',
      host: 'Browser',
      action: 'open',
      param: param,
    });
 */
function getProtocol() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  // 参数默认值
  options.param.success = options.param.success || function () {};
  options.param.error = options.param.error || function () {};
  // 回调名称
  var callbackName = 'native_callback_' + Math.random().toString().split('.')[1];
  if (_dvdBaseJsRuntime2.default.isClient()) {

    // 设置全局回调
    window[callbackName] = function (response) {
      // 多数native接口执行成功时response.code就会返回'1',执行失败时就会返回'0'。
      // 此处返回'1'时候正常回调success,但返回不等于'1'时一律执行error。
      // 如此一来,诸如Account.login这样的返回'0'|'1'|'2'的接口就可以处理'0'|'1'以外的情况了(在error中判断response === '0'|'2'|'3'|'4'|...)。
      if (response.code == '1') {
        if (options.param && options.param.success) options.param.success(response);
        if (options.param && options.param.content && options.param.content.success) {
          options.param.content.success(response);
        }
      } else {
        options.param.error(response);
      }
      // 执行完回收
      window[callbackName] = null;
    };

    // 设置分享回调
    if (options.param.shareCallBack) {
      var shareCb = options.param.shareCallBack;
      var shareCbName = 'native_share_callback_' + Math.random().toString().split('.')[1];
      window[shareCbName] = function (response) {
        // 执行业务回调
        shareCb(response);
      };
      options.param.shareCallBack = shareCbName;
    }
  }

  // 拼接唤起native协议
  var protocol = 'davdian://call.' + options.host + '.com?action=' + options.action + '&params=' + encodeURIComponent(JSON.stringify(options.param)) + '&callback=' + callbackName + '&minv=' + options.v;
  return protocol;
}

/**
 * 功能: 端内调用app
 * 用法:
 * call({
      v: '3.1.0',
      host: 'Browser',
      action: 'open',
      param: {},
    });
 */
function innerCall() {
  var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  // 不在app中,直接返回
  if (!_dvdBaseJsUa2.default.isDvdApp()) return;
  // 参数检查
  if (!param.host || !param.action || !param.v) {
    throw new Error('参数不全');
  }
  // 参数默认值
  param.invalid = param.invalid || function () {
    // popup.toast("请升级您的APP到最新版");
  };

  // 当前版本高于指定版本,可以调用native,否则进行提示
  if (_dvdBaseJsUa2.default.compareVersion(_dvdBaseJsUa2.default.getDvdAppVersion(), param.v) >= 0) {
    // 客户端端内调用协议
    var protocol = getProtocol(param);

    // 是否输出调试信息
    if (param.debug) {
      // 在页面底部打印调用native协议
      var logDom = document.createElement('div');
      logDom.style.margin = '10px';
      logDom.textContent = protocol;
      document.body.appendChild(logDom);
      _dvdServiceJsPopup2.default.debug({
        title: protocol
      });
      alert(protocol);
    }

    // 日志
    console.log(protocol);

    // 调用native, timeout解决协议覆盖问题
    // location.href = protocol;

    // 调用native
    var iframe = document.createElement('iframe');
    iframe.style.display = 'none';
    iframe.src = protocol;
    document.body.appendChild(iframe);
  } else {
    // 版本错误提示
    param.invalid();
  }
}

/**
 * 功能: 端外调用app(唤起后再执行app接口)
 * 用法:
 * call({
      v: '3.1.0',
      host: 'Browser',
      action: 'open',
      param: {
        invoke: true,
        ...
      },
    });
 */
function outerCall() {
  var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var stayCurrentPage = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

  // 在app中,直接返回
  if (_dvdBaseJsUa2.default.isDvdApp()) return;

  // 端内调用协议
  var protocol = getProtocol(param);

  if (_dvdBaseJsUa2.default.isAndroid()) {
    // 端外唤起协议
    var outerProtocol = 'davdian://invoke.davdian.com?cmd=' + encodeURIComponent(protocol);

    // 打印协议
    console.log(outerProtocol);

    if (_dvdBaseJsRuntime2.default.isClient()) {
      // 唤起app
      location.href = outerProtocol;

      if (!stayCurrentPage) {
        // 兜底跳转下载页
        setTimeout(function () {
          location.href = '//a.app.qq.com/o/simple.jsp?pkgname=com.davdian.seller';
        }, 3000);
      }
    }
  } else if (_dvdBaseJsUa2.default.isIos()) {
    // let env = /(davdian\.com|vyohui\.cn|bravetime\.net)/.exec(location.href)[0];

    // 端外唤起协议
    var _outerProtocol = '//invoke.davdian.com?cmd=' + encodeURIComponent(protocol);

    // 打印协议
    console.log(_outerProtocol);

    // alert(protocol)
    // alert(outerProtocol);

    if (_dvdBaseJsRuntime2.default.isClient()) {
      // 唤起app
      location.href = _outerProtocol;
    }
  }
}

/**
 * 功能: 调用native(核心方法)
 * 用法:
 * call({
      v: '3.1.0',
      host: 'Browser',
      action: 'open',
      param: {
        invoke: true,
        ...
      },
    });
 * stayCurrentPage 为true 代表从外部跳转时不跳转应用宝 留在当前页
 */
function call() {
  var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var stayCurrentPage = arguments[1];

  // 如果param.param.invoke为真,需要先唤起native再执行cmd
  var a = '';
  for (var p in param.param.content) {
    a += p + '=' + param.param.content[p];
  }
  // alert(a);
  if (param.param.invoke) {
    delete param.param.invoke;
    outerCall(param, stayCurrentPage);
  } else {
    innerCall(param);
  }
}

var native = {
  /****************************** 1、账户相关 ******************************/
  Account: {
    /**
     * 功能: 登录
     * 用法:
     * native.Account.login()
     */
    login: function login() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var debug = arguments[1];

      setTimeout(function () {
        call({
          v: '2.4.0',
          host: 'Account',
          action: 'login',
          param: param,
          debug: debug
        });
      }, 500);
    }
  },

  /****************************** 2、直播相关 ******************************/
  LiveVideo: (_LiveVideo = {
    /**
     * 功能: 跳转直播只有当App未启动，或处于首页4个tab页面时（即首页上未盖住其他页面的情况）才进行跳转，否则不进行跳转，以防止出现无法预期的异常情况
     * 用法:
     * native.LiveVideo.goTab({
        userId: "用户Id",
      })
     */
    goTab: function goTab() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.4.0',
        host: 'LiveVideo',
        action: 'goTab',
        param: param
      });
    },


    /**
     * 功能: 进入某个直播间isPlaying决定客户端在获取到状态为回顾时，是否需要提示；fromPush用来决定如何提示用户（推送和普通流程不同）
     * 用法:
     * native.LiveVideo.enterRoom({
       "liveId" : "房间号",
       "isPlaying":"1/2/3" ,// 1表示直播中，2是回放，3是整理中
       "fromPush":"0/1" // 0表示不来自于推送，1表示来自推送
      })
     */
    enterRoom: function enterRoom() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.5.0',
        host: 'LiveVideo',
        action: 'enterRoom',
        param: param
      });
    },


    /**
     * 功能: 进入直播申请页，若已有进行中申请，则显示申请进度（审核中或审核被拒绝）
     * 用法:
     * native.LiveVideo.applyVideoLive()
     */
    applyVideoLive: function applyVideoLive() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.5.0',
        host: 'LiveVideo',
        action: 'applyVideoLive',
        param: param
      });
    },


    /**
     * 功能: 展示直播申请结果
     * 用法:
     * native.LiveVideo.showApplyResult({
        "result":"0/1" // 0表示未通过，1表示通过
      })
     */
    showApplyResult: function showApplyResult() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.5.0',
        host: 'LiveVideo',
        action: 'showApplyResult',
        param: param
      });
    },


    /**
     * 功能: 进入某个话题列表
     * 用法:
     * native.LiveVideo.openTopicVideoList({
        "type":"video/voice",//video:视屏直播，voice:语音回顾
        "name":"话题标题" ,
        "ID":"话题ID"
      })
     */
    openTopicVideoList: function openTopicVideoList() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.7.0',
        host: 'LiveVideo',
        action: 'openTopicVideoList',
        param: param
      });
    },


    /**
     * 功能: 进入全部话题列表
     * 用法:
     * native.LiveVideo.openTopicList()
     */
    openTopicList: function openTopicList() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.7.0',
        host: 'LiveVideo',
        action: 'openTopicList',
        param: param
      });
    }
  }, _defineProperty(_LiveVideo, 'openTopicVideoList', function openTopicVideoList() {
    var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    call({
      v: '2.7.0',
      host: 'LiveVideo',
      action: 'openTopicVideoList',
      param: param
    });
  }), _defineProperty(_LiveVideo, 'openEssenceList', function openEssenceList() {
    var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    call({
      v: '3.4.0',
      host: 'LiveVideo',
      action: 'openEssenceList',
      param: param
    });
  }), _LiveVideo),

  /****************************** 3、用户信息 ******************************/
  PersonInfo: {
    /**
     * 功能: 查看(大V课讲师的)个人主页
     * 用法:
     * native.PersonInfo.showPersonHome({
        userId: "用户Id",
      })
     */
    showPersonHome: function showPersonHome() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.5.0',
        host: 'PersonInfo',
        action: 'showHomePage',
        param: param
      });
    }
  },

  /****************************** 4、浏览 ******************************/
  Browser: {
    /**
     * 功能: 打开新的webview
     * 用法:
     * native.Browser.open({
        url: 'http://18514008282.vyohui.cn/',
      });
     */
    open: function open() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.5.0',
        host: 'Browser',
        action: 'open',
        param: param
      });
    },


    /**
     * 功能: 关闭当前webview
     * 用法:
     * native.Browser.close();
     */
    close: function close() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.5.0',
        host: 'Browser',
        action: 'close',
        param: param
      });
    },


    /**
     * 功能: 设置native头部标题栏
     * 用法:
     * native.Browser.setHead({
        'title' : '修改后的标题',
        'backBtn' : '0/1', // 0表示头部不展示返回按钮，1表示展示
        'homeBtn' : '0/1', // 0表示头部不展示首页按钮，1表示展示
        'shareBtn' : '0/1', //0表示头部不展示分享按钮，1表示展示
        'shareMoney' : '分享获得返现数', //0表示分享返现为0不显示分享赚钱按钮，非0：展示分享赚钱按钮，在3.9.1之前用该字段
        'shareMoneyStr':'分享获取返现数显示的字符串',//在3.9.1中会用该字段
        'rightBtn' : {
          'text' : '头部右部展示的文字按钮文字',        
          'textColor':'头部右部按钮文字颜色',
          'action' : '文字按钮需要执行的JS'
        },
        'hideHead' : '0',
      })
     */
    setHead: function setHead() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var debug = arguments[1];

      call({
        v: '2.6.0',
        host: 'Browser',
        action: 'setHead',
        param: param,
        debug: debug
      });
    },


    /**
     * 功能: 初始化native头部标题栏
     * 用法:
     * native.Browser.initHead({
        content: JSON.stringify({
          showHead: 1,
          showFoot: 0,
          backOnHead: 1,
          homeOnHead: 0,
          shareOnHead: 0,
          btnOnHead: 0,
          btnText: "",
          btnLink: ""
        })
     })
     */
    initHead: function initHead() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.4.0',
        host: 'Browser',
        action: 'initHead',
        param: param
      });
    },


    /**
     * app支付
     * 用法:
     * native.Browser.pay({
        "url": "",
        "pay_type": "1",   // payType的值为0则为默认1为组团参与者，2为组团发起者。
        "openDVD": "0",     // 判断是否为开店支付：0为普通支付，1为开店支付  该参数在4.2.0之后用到
        "pay_id": "xxx",    // 支付ID
        "sign": "",
        "platform": "2",     // 直接调用支付平台，0：直接调用微信支付，1：直接调用支付宝支付。2:苹果内购
        "appleGoods_id": '',  // 苹果的商品ID(只在内购中使用)
        "iap_type": ''        // 内购的商品类型,取值范围：album(合辑音频类型)， course(大V课)，vip(购买会员)
      });
     */
    pay: function pay() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.1.0',
        host: 'Browser',
        action: 'pay',
        param: param
      });
    },


    /**
     * 功能: 添加订单选择身份证
     * 用法:
     * native.Browser.selectIdentity({
        cardName: ''
      })
     */
    selectIdentity: function selectIdentity() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.7.0',
        host: 'Browser',
        action: 'selectIdentity',
        param: param
      });
    },


    /**
     * 功能: APP中点击图片显示大图
     * 用法:
     * native.Browser.selectIdentity({
        cardName: ''
      })
     */
    showBigImage: function showBigImage() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '4.1.0',
        host: 'Browser',
        action: 'showBigImage',
        param: param
      });
    },
    showWebHeight: function showWebHeight() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '4.2.0',
        host: 'Browser',
        action: 'showWebHeight',
        param: param
      });
    },
    goNativeHomePage: function goNativeHomePage() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '4.2.0',
        host: 'Browser',
        action: 'goNativeHomePage',
        param: param
      });
    },


    /**
     * 1018爆款商品预约
     * native.Browser.goodsBook({
        goodsId: '商品Id',
        goodsTitle: '商品标题',
        goodsImage: '商品图片',
        goodsStartTime: '开抢时间',
        goodsUrl: '商品详情链接',
        goodsListUrl: '商品列表链接'
      });
     */
    goodsBook: function goodsBook() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var debug = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      call({
        v: '5.1.0',
        host: 'Browser',
        action: 'goodsBook',
        param: param,
        debug: debug
      });
    },


    /**
     * 返回上一页
     * native.Browser.goBack({
        needRefresh : 0  // 是否需要刷新 1需要，0不需要，默认为不需要
      });
     */
    goBack: function goBack() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.2.0',
        host: 'Browser',
        action: 'goBack',
        param: param
      });
    },


    /**
     * 大v店金币苹果充值
     * native.Browser.payRecharge({
        "appleGoods_id": '', //苹果的商品ID(只在内购中使用)
        "priceId":"2234"
      });
     */
    payRecharge: function payRecharge() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'Browser',
        action: 'payRecharge',
        param: param
      });
    },


    /**
     * 使用苹果内购购买会员
     * native.Browser.payVip({
         "appleGoods_id": '', //苹果的商品ID(只在内购中使用)
         "dvdItemId":"2234"
      });
     */
    payVip: function payVip() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'Browser',
        action: 'payVip',
        param: param
      });
    },


    /**
     * 返回上一页
     * native.Browser.openImageBrowser({
         currentPageIndex: "1"  //"当前数据的下一页页码（当前为第2页数据则传3）"
         id: "1"  //"id"
         currentImgUrl: "http://asdsadsadd.jpg"  //"用户点击的图片地址"
         jsonString:   //"json数据，格式为string类型" 格式：{data:[{"comment":"评论内容","imageList":["imageUrl",..]},..]}
     });
     */
    openImageBrowser: function openImageBrowser() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.5.0',
        host: 'Browser',
        action: 'openImageBrowser',
        param: param
      });
    },


    /**
     * 获取地理位置
     * native.Browser.getLocation({
        "code" : "0/1" // 1表示返回成功，0表示返回失败
        "country" : "中国"  //国家
        "state"："北京"     //省
        "city":"北京"         //市
        "district":"朝阳区"  //区
        "address":"中国北京市朝阳区望京东街1号摩托罗拉大厦"
        "location":"123.342342,23.234234"  //坐标
     });
     */
    getLocation: function getLocation() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Browser',
        action: 'getLocation',
        param: param
      });
    },


    /**
     * 图片上传
     * native.Browser.uploadImage({
       "sourceType" : "0/1/2",//0表示相机拍摄；1表示相册选择；2表示弹出选择相机拍照和相册选择的框；
       "maxImageNumber": "9",   //允许一次上传的图片的最大数限制,最大设置9张;
       "compress":"0到1",//质量压缩比例 1为不压缩  0.8为压缩至原图的80%
       "tailor"："0/1",//是否要求裁剪，0,不裁剪，1，裁剪；只有单张图支持裁剪，多张图不支持裁剪；
       "biz":"业务参数"//1、新鲜事 2：商品 3：我的 4：学院直播 5：直播缩略图 6：故事妈妈
     });
     */
    uploadImage: function uploadImage() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Browser',
        action: 'uploadImage',
        param: param
      });
    }
  },

  /****************************** 5、购物车 ******************************/
  ShopCart: {
    /**
     * 功能: 商品加入购物车
     * 用法:
     * native.ShopCart.open({
        "goodsID" : "商品ID",
        "cout" : "商品数量"
      });
     */
    addGoodsToShopCart: function addGoodsToShopCart() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '2.6.0',
        host: 'ShopCart',
        action: 'addGoodsToShopCart',
        param: param
      });
    }
  },

  /****************************** 6、发现社区 ******************************/
  Community: {
    /**
     * 功能: 打开发现社区
     * 用法:
     * native.Community.openCommunity(;
     */
    openCommunity: function openCommunity() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.1.0',
        host: 'Community',
        action: 'openCommunity',
        param: param
      });
    }
  },

  /****************************** 7、share分享 ******************************/
  Share: {
    /**
     * 分享卡
     * 用法:
     * native.Share.shareInfo({
        "show":"是否显示下方shareView",//0：不显示，1：显示。
        "shareType":"是否只分享图片"，//0：分享图文链接方式，1：只分享图片 （支持图片分享的分享图片、不支持分享图文链接（需要传tilte、desc、link、imgUrl））。
        "sharePlatform":"分享方式",//0：微信朋友圈，1：微信好友，2：QQ好友，3：QQ空间，4：短信，5：复制功能。
        "shareTitle":"底部特殊分享标题",
        "shareDesc":"底部特殊分享描述",
        "title":"分享标题",
        "desc":"分享内容描述",
        "link":"分享链接',
        "imgUrl":"分享图片链接",
        "bigImageUrl":"只分享图片：大图片链接",
        "shareSource":"分享统计的shareSource", //h5分享为15或者没有这个字段，原生的一定要传根据分享列表传值http://wiki.bravetime.net/pages/viewpage.action?pageId=15106888
      })
     */
    shareInfo: function shareInfo() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: param.v || '3.3.0',
        host: 'Share',
        action: 'shareInfo',
        param: param
      });
    },


    /**
     * 分享卡
     * 用法:
     * native.Share.share({
        courseId: '直播ID',
      })
     */
    shareCard: function shareCard() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.4.0',
        host: 'Share',
        action: 'cardShare',
        param: param
      });
    },


    /**
     * 素材分享
     * 用法:
     * native.Share.materialShare({
        "pageId":"素材分享页id"
      })
     */
    materialShare: function materialShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.6.0',
        host: 'Share',
        action: 'materialShare',
        param: param
      });
    },


    /**
     * 小书库分享卡分享
     * 用法:
     * native.Share.recordCardShare({
        "recordId":"打卡内容Id"
     })
     */
    recordCardShare: function recordCardShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'Share',
        action: 'recordCardShare',
        param: param
      });
    },


    /**
     * 海报分享卡分享
     * 用法:
     * native.Share.posterCardShare({
        "recordId":"打卡内容Id"
     })
     */
    posterCardShare: function posterCardShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.5.0',
        host: 'Share',
        action: 'posterCardShare',
        param: param
      });
    },


    /**
     * 单平台分享
     * 用法:
     * native.Share.singlePlatformShare({
       "shareType":"分享的内容样式"，//0：分享图文链接方式，1：只分享图片 2:只分享文字 （没有此值默认分享图文链接）
        "sharePlatform":"显示的分享平台",//0：微信朋友圈，1：微信好友，2：QQ好友，3：QQ空间，4：复制功能。
        "title":"分享标题",
        "desc":"分享内容描述",
        "link":"分享链接',
        "imgUrl":"分享图片Url"
      })
     */
    singlePlatformShare: function singlePlatformShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.5.0',
        host: 'Share',
        action: 'singlePlatformShare',
        param: param
      });
    },


    /**
     * 多图分享
     * 用法:
     * native.Share.imgListShare({
        //分享信息
        text: "分享文案",
        imgList: ["http://asdasd.png", "http://dsadwdqfqwf.png"], //json字符串
        //打点统计信息
        production: "31", //来源页面，31代表商品详情页面
        actionType: "0/1/2", // 页面点击的位置 0：默认，1：商品详情页右上角分享按钮点击调起面板 2:商品详情页一键分享按钮点击调起面板
      })
     */
    imgListShare: function imgListShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Share',
        action: 'imgListShare',
        param: param
      });
    },


    /**
     * 视频分享
     * 用法:
     * native.Share.videoShare({
        //分享信息
        imgUrl: "http://sadasdasd.png", //分享的图标
        title: "分享标题",
        text: "分享文案",
        url: "跳转链接",
        videoUrl: "http://asdisadniaondosand.mp4", //视频地址
        //打点统计信息
        production: "31", //来源页面，31代表商品详情页面
        actionType: "0/1/2", // 页面点击的位置 0：默认，1：商品详情页右上角分享按钮点击调起面板 2:商品详情页一键分享按钮点击调起面板
      })
     */
    videoShare: function videoShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Share',
        action: 'videoShare',
        param: param
      });
    },


    /**
     * 商品素材分享
     * 用法:
     * native.Share.goodsMaterialShare({
        //图文链接分享信息
        "title": "面板的标题文案",
        "desc": "面板的描述文案",
        "imgUrl": "http://sadasdasd.png", //图文链接分享的图片地址
        "shareTitle": "图文链接的分享标题",
        "shareDesc": "图文链接的分享描述",
        "url": "跳转地址",
        //素材分享信息
        "goodsImageUrl": "http://pic.jj.jpg", //h5合成的图片二维码地址
        "materials": [{
          "materialType": "1"//1表示图片，2表示视频,
          "copyWriting": "商品素材文案",
          "images": [
            "http://pic.davdian.com/supplier/2017/08/14/800_800_a906cb3d1bb9013661ccf3f977704913.jpg", //图片地址，当素材类型为1时才会有值},
            {
              "materialType": 2,//素材类型1、图片；2、视频
              "copyWriting": "商品素材文案",//文案
              //图片地址，当素材类型为1时才会有值
              "videoUrl": "http://media.davdian.com/goodsVideo/2017/06/14/ad614c02df201d91a96e1e9e2edf2177.mp4",//视频地址
              "videoImg": "http://pic.davdian.com/supplier/2017/08/14/800_800_a906cb3d1bb9013661ccf3f977704913.jpg",//视频图片地址
            },
          ],
          //打点统计信息
          "production": "31", //来源页面，31代表商品详情页面
          "actionType": "0/1/2", // 页面点击的位置 0：默认，1：商品详情页右上角分享按钮点击调起面板 2:商品详情页一键分享按钮点击调起面板}
     }],
     })
     */
    goodsMaterialShare: function goodsMaterialShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Share',
        action: 'goodsMaterialShare',
        param: param
      });
    },


    /**
     * 单图分享
     * 用法:
     * native.Share.imageShare({
        //分享信息
        "imgUrl": "http://sadasdasd.png", //分享图片url
        //打点统计信息
        "production": "31", //来源页面，31代表商品详情页面
        "actionType": "0/1/2", // 页面点击的位置 0：默认，1：商品详情页右上角分享按钮点击调起面板 2:商品详情页一键分享按钮点击调起面板
      });
     */
    imageShare: function imageShare() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Share',
        action: 'imageShare',
        param: param
      });
    }
  },

  /****************************** 8、语音直播相关 ******************************/
  VoiceLive: {
    /**
     * 功能: 打开语音直播介绍页
     * 用法:
     * native.VoiceLive.openVoiceIntro({
        "courseId":"语音直播间id",//语音直播间id，根据courseId去请求得到加载网址。
        "fromPush":"判断是否未推送",//0:代表非推送打开，1：推送打开 。
        "messageType":"推送消息类型",//只有判断为推送消息才会解析该参数 1：语音直播即将开始消息，2:语音直播问题被回答消息；
      });
     */
    openVoiceIntro: function openVoiceIntro() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.4.0',
        host: 'VoiceLive',
        action: 'openVoiceIntro',
        param: param
      });
    },


    /**
     * 功能: 打开老师信息页
     * 用法:
     * native.VoiceLive.openTeacherHome({
        "teacherId" : "老师ID"
      });
     */
    openTeacherHome: function openTeacherHome() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.4.0',
        host: 'VoiceLive',
        action: 'openTeacherHome',
        param: param
      });
    },


    /**
     * 功能: 进入直播间
     * 用法:
     * native.VoiceLive.enterVoiceRoom({
        "courseId": "语音直播间id"
      });
     */
    enterVoiceRoom: function enterVoiceRoom() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.4.0',
        host: 'VoiceLive',
        action: 'enterRoom',
        param: param
      });
    },


    /**
     * 功能: 进入全部课程列表
     * 用法:
     * native.VoiceLive.enterCourseList();
     */
    enterCourseList: function enterCourseList() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.7.0',
        host: 'VoiceLive',
        action: 'enterCourseList',
        param: param
      });
    },


    /**
     * 功能: 进入语音直播
     * 用法:
     * native.VoiceLive.enterOldVoiceLiveRoom({
         "liveId": "149",//直播id
         "title": "著名亲子教育专家 卢洪波：孩子贪玩，我该怎么办？",//标题
         "status": "4",//状态（2：直播结束前 3：直播整理中 4：直播回顾）
         "intro": "分享时间:2016年07月18日（周一）14:00-15:00　　分享内容:　　很多父母一直处于困惑纠结中，孩子小的时候，父母想尽一切办法教TA玩，但是孩子从上幼儿园开始，父母开始限制TA玩，上小学后，几乎就不允许玩，然后就会发现孩子是想尽一切办法在找时间偷着玩：听课时，因为没有玩具，玩手指头；写作业时，趁着思考的间隙，玩转笔；走路时，没有足球，踢石头子......家长挂在嘴边上的话又多了一句“别玩，好好学习”“别玩，快写！”“别玩，好好走路”等等。　　　　作为父母，我们是怎么认识玩的呢？孩子的成长是否需要玩呢？到底应该让孩子玩什么，怎么玩呢？在玩中，应该培养哪些优秀好好学习”“别玩，快写！”“别玩，好好走路”等等。　　　　作为父母，我们是怎么认识玩的呢？孩子的成长是否需要玩呢？到底应该让孩子玩什么，怎么玩呢？在玩中，应该培养哪些优秀\345\223品质呢？如何实现在玩中学，学中玩的理想教育呢？PS:&nbsp;&nbsp;在大V店APP里预约听课，可以立即向分享老师提问哦！听课指南："(需要将/n、/r替换为//n、//r),//课程简介、分享描述内容
         "smallImageUrl": "http://pic.davdian.com/free/2016/07/01/320_320_d9c182776612fe1f3243b763dcf9a1b5.jpg",//分享显示图片
         "reviewUrl": "/index.php?m=default&c=live&a=liveinfo&id=149",//回顾网页
         "shareUrl": "/index.php?m=default&c=live&a=liveinfo&id=149"//分享的链接
       });
     */
    enterOldVoiceLiveRoom: function enterOldVoiceLiveRoom() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.7.0',
        host: 'VoiceLive',
        action: 'enterOldVoiceLiveRoom',
        param: param
      });
    },

    /**
     * 功能: 进入全部笔记页面
     * 用法:
     * native.VoiceLive.callAppEnterAllNote()
     });
     */
    callAppEnterAllNote: function callAppEnterAllNote() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.9.2',
        host: 'VoiceLive',
        action: 'openCourseCommentList',
        param: param
      });
    },

    /**
     * 功能: 进入写笔记页面
     * 用法:
     * native.VoiceLive.callAppEnterWriteNote()
     });
     */
    callAppEnterWriteNote: function callAppEnterWriteNote() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '4.1.0',
        host: 'VoiceLive',
        action: 'openCourseNoteEdit',
        param: param
      });
    }
  },

  /****************************** 9、社群 ******************************/
  CommunityChat: {
    /**
     * 功能: 打开社群
     * 用法:
     * native.CommunityChat.enterCommunityChat();
     */
    enterCommunityChat: function enterCommunityChat() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.8.0',
        host: 'CommunityChat',
        action: 'enterCommunityChat',
        param: param
      });
    },


    /**
     * 功能: 打开社群成员列表
     * 用法:
     * native.CommunityChat.enterCommunityMemberList({
        "communityId" : "社群id"
      });
     */
    enterCommunityMemberList: function enterCommunityMemberList() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.8.0',
        host: 'CommunityChat',
        action: 'enterCommunityMemberList',
        param: param
      });
    }
  },

  /****************************** 10、搜索相关 ******************************/
  Search: {
    /**
     * 功能: 进入商品搜索结果页
     * 用法:
     * native.Search.enterGoodsSearchResult({
        "name" : "书包"
      });
     */
    enterGoodsSearchResult: function enterGoodsSearchResult() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.9.0',
        host: 'Search',
        action: 'enterGoodsSearchResult',
        param: param
      });
    }
  },

  /****************************** 11、H5触发cmd命令 ******************************/
  BrowserTouch: {
    /**
     * 功能: H5触发cmd客户端实现回到上一级页面
     * 用法:
     * native.BrowserTouch.goBackToRootPage();
     */
    goBackToRootPage: function goBackToRootPage() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '3.9.0',
        host: 'BrowserTouch',
        action: 'goBackToRootPage',
        param: param
      });
    },

    /**
     * 功能: H5触发cmd客户端复制文案到剪切板
     * 用法:
     * native.BrowserTouch.copyText();
     */
    copyText: function copyText() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '4.2.0',
        host: 'BrowserTouch',
        action: 'copyText',
        param: param
      });
    },

    /**
     * 功能: 大V课显示课程编辑
     * 用法:
     * native.BrowserTouch.showCourseEdit({
        marginRight: '10',
     });
     */
    showCourseEdit: function showCourseEdit() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.2.0',
        host: 'BrowserTouch',
        action: 'showCourseEdit',
        param: param
      });
    }
  },
  /****************************** 12、H5触发cmd命令 ******************************/
  Common: {
    /**
     * 功能: H5触发cmd客户端实现回到上一级页面
     * 用法:
     * native.Common.log();
     */
    log: function log() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '4.1.0',
        host: 'Common',
        action: 'log',
        param: param
      });
    },


    /**
     * 下载图片集合
     * 用法:
     * native.Common.downloadImages({
        "imgList": [
          "http://asdasdiasndiad.png",
          "http://asdasdiasndiad.png"
        ]
      });
     */
    downloadImages: function downloadImages() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Common',
        action: 'downloadImages',
        param: param
      });
    },


    /**
     * 下载视频
     * 用法:
     * native.Common.downloadVideo({
        "videoUrl": "http://asdhisaodioasd.mp4"
      });
     */
    downloadVideo: function downloadVideo() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.6.0',
        host: 'Common',
        action: 'downloadVideo',
        param: param
      });
    }
  },

  /****************************** 13、音频command命令 ****************************/
  Audio: {
    /**
     * 功能: H5触发cmd客户端实现回到上一级页面
     * 用法:
     * native.Audio.goAudioDetail();
     */
    goAudioDetail: function goAudioDetail() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.0.0',
        host: 'Audio',
        action: 'goAudioDetail',
        param: param
      });
    },

    /**
     * 功能: H5触发cmd客户端播放暂停音频
     * 用法:
     * native.Audio.audioPlay(
     *  {
          "sortNo" : "57",//当前被点击的音频下标索引
          "alummId" : "13",//当前合辑id
          "soundId" : "声音Id", // 音频id
         }
     * );
     */
    audioPlay: function audioPlay() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      if (param.sortNo) {
        param.sortNo = param.sortNo.toString();
      }
      param.albumId = param.albumId.toString();
      call({
        v: '5.0.0',
        host: 'Audio',
        action: 'audioPlay',
        param: param
      });
    },

    /**
     * 功能: H5触发cmd客户端回调js告诉我们正在播放是哪一条
     * 用法:
     * native.Audio.audioLocation();
     */
    audioLocation: function audioLocation() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.0.0',
        host: 'Audio',
        action: 'audioLocation',
        param: param
      });
    },

    /**
     * 功能: H5触发cmd客户端回调js告诉我们上次听到了那一条
     * 用法:
     * native.Audio.audioLocation();
     */
    audioPlayHistory: function audioPlayHistory() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.0.0',
        host: 'Audio',
        action: 'audioPlayHistory',
        param: param
      });
    },

    /**
     * 功能:H5触发cmd客户端订阅某个音频
     * 用法:
     * native.Audio.audioSubscription();
     */
    audioSubscription: function audioSubscription() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.0.0',
        host: 'Audio',
        action: 'audioSubscription',
        param: param
      });
    }
  },

  /****************************** 14、打开二维码扫码页面command命令 ****************************/
  ScanCode: {
    /**
     * 功能: 打开二维码
     * 用法:
     * native.ScanCode.openScanCode();
     */
    openScanCode: function openScanCode() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.1.0',
        host: 'ScanCode',
        action: 'openScanCode',
        param: param
      });
    }
  },

  /****************************** 15、小书库command命令 ****************************/
  BookStore: {
    openActiveDetail: function openActiveDetail() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'openActiveDetail',
        param: param
      });
    },
    openContentDetail: function openContentDetail() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'openContentDetail',
        param: param
      });
    },
    openCustomRecord: function openCustomRecord() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'openCustomRecord',
        param: param
      });
    },
    enterPerusalList: function enterPerusalList() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'enterPerusalList',
        param: param
      });
    },
    enterPerusal: function enterPerusal() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'enterPerusal',
        param: param
      });
    },
    thumbUp: function thumbUp() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'thumbUp',
        param: param
      });
    },
    imagesBrowser: function imagesBrowser() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.3.0',
        host: 'BookStore',
        action: 'imagesBrowser',
        param: param
      });
    }
  },

  /****************************** 16、混合开发command命令 ****************************/
  AppHybrid: {
    memberPage: function memberPage() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.7.0',
        host: 'AppHybrid',
        action: 'memberPage',
        param: param
      });
    },
    hidePlaceholderImage: function hidePlaceholderImage() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      call({
        v: '5.7.0',
        host: 'AppHybrid',
        action: 'hidePlaceholderImage',
        param: param
      });
    }
  },

  /****************************** 自定义便捷接口 ******************************/
  custom: {
    /**
     * 功能: 设置native分享信息
     * 用法:
     * native.custom.setShareInfo({
        title: '大V店组团包邮', // 分享标题
        desc: '一件包邮！每天上新！好货低价又包邮，抢到了就赚翻啦', // 分享描述
        link: location.href, // 分享链接
        imgUrl: 'http://pic.davdian.com/free/2016/04/09/320_320_0fc3e0dbbadd249b7f1b93a525f0adf0.jpg', // 分享图标
      });
     */
    setShareInfo: function setShareInfo() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 参数合并
      param = _jquery2.default.extend({}, config.defaultShareInfo, param);

      // 参数处理
      param.imgUrl.replace("https", "http");

      if (_dvdBaseJsRuntime2.default.isClient()) {
        // 注册全局方法供native调用
        window.iosInterface = window.iosInterface || {};
        window.iosInterface.getShareInfo = function () {
          return JSON.stringify(param);
        };
      }
    },


    /**
     * 功能: 注册事件回调,webview返回当前页面时执行。
     * 背景: ios3.9.1之后采用WKWebView,返回当前页面时不会自动刷新页面,但ios会回调js全局方法。
     * 用法:
     * native.custom.onWebviewBack({
        callback: function () {
          location.reload();
        }
      });
     */
    onWebviewBack: function onWebviewBack() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      if (_dvdBaseJsRuntime2.default.isClient()) {
        // 参数处理
        param.callback = param.callback || function () {
          location.reload();
        };
      }

      if (typeof window !== 'undefined') {
        // 注册全局方法供native调用
        window.iosInterface = window.iosInterface || {};
        var nativeWebviewBack = window.iosInterface.nativeWebviewBack;
        window.iosInterface.nativeWebviewBack = function () {
          nativeWebviewBack && nativeWebviewBack();
          param.callback();
        };
      }
    },


    /**
     * 功能: 初始化native头部标题栏
     * 用法:
     * native.custom.initHead({
        showHead: 1,
        showFoot: 0,
        backOnHead: 1,
        homeOnHead: 0,
        shareOnHead: 0,
        btnOnHead: 0,
        btnText: "",
        btnLink: "",
      })
     */
    initHead: function initHead() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 参数合并
      param = _jquery2.default.extend({}, defaultInitHead, param);
      // 调用Browser.initHead接口
      native.Browser.initHead({
        content: param
      });
    },


    /**
     * 功能: 初始化native头部标题栏
     * 用法:
     * native.custom.setHead({
        'showHead': '1',
        'title' : '修改后的标题',
        'backBtn' : '0', // 0表示头部不展示返回按钮，1表示展示
        'homeBtn' : '0', // 0表示头部不展示首页按钮，1表示展示
        'shareBtn' : '0', //0表示头部不展示分享按钮，1表示展示
        'shareMoney' : '分享获得返现数', //0表示分享返现为0不显示分享赚钱按钮，非0：展示分享赚钱按钮，在3.9.1之前用该字段
        'shareMoneyStr':'分享获取返现数显示的字符串',//在3.9.1中会用该字段
        'rightBtn' : {
          'text' : '头部右部展示的文字按钮文字',        
          'textColor':'头部右部按钮文字颜色',
          'action' : '文字按钮需要执行的JS'
        },
        'hideHead' : '0',
      })
     */
    setHead: function setHead() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var debug = arguments[1];

      // 参数合并
      param = _jquery2.default.extend({}, defaultSetHead, param);

      // 调用Browser.setHead接口
      native.Browser.setHead(param, debug);
    },


    /**
     * 功能: 分享
     * 用法:
     * native.custom.share({
        title: 'MAMA+|大V店',
        desc: 'MAMA+|大V店',
        link: 'MAMA+|大V店',
        imgUrl: 'http://pic.davdian.com/goods/1/20151017104524.png'
      })
     */
    share: function share() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      if (_dvdBaseJsRuntime2.default.isClient()) {
        // 分享信息(页面分享信息+默认分享信息)
        var shareInfo = _jquery2.default.extend({}, config.defaultShareInfo, window.iosInterface && window.iosInterface.getShareInfo && JSON.parse(window.iosInterface.getShareInfo()));

        // 默认参数(加入分享信息)
        var defaultParam = _jquery2.default.extend({
          show: '1',
          shareType: '0'
        }, shareInfo);

        // 覆盖默认参数
        param = _jquery2.default.extend({}, defaultParam, param);

        // 调用Share.shareInfo接口
        native.Share.shareInfo(param);
      }
    },


    /**
     * 功能: 分享朋友圈
     * 用法:
     * native.custom.shareWeixinTimeline({
        title: 'MAMA+|大V店',
        desc: 'MAMA+|大V店',
        link: 'MAMA+|大V店',
        imgUrl: 'http://pic.davdian.com/goods/1/20151017104524.png'
      })
     */
    shareWeixinTimeline: function shareWeixinTimeline() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '0',
        sharePlatform: '0'
      }, param);
      this.share(param);
    },


    /**
     * 分享微信好友
     * 用法:
     * native.custom.shareWeixinFriend({
        title: 'MAMA+|大V店',
        desc: 'MAMA+|大V店',
        link: 'MAMA+|大V店',
        imgUrl: 'http://pic.davdian.com/goods/1/20151017104524.png'
      })
     */
    shareWeixinFriend: function shareWeixinFriend() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '0',
        sharePlatform: '1'
      }, param);
      this.share(param);
    },


    /**
     * 分享QQ好友
     * 用法:
     * native.custom.shareQQFriend({
        title: 'MAMA+|大V店',
        desc: 'MAMA+|大V店',
        link: 'MAMA+|大V店',
        imgUrl: 'http://pic.davdian.com/goods/1/20151017104524.png'
      })
     */
    shareQQFriend: function shareQQFriend() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '0',
        sharePlatform: '2'
      }, param);
      this.share(param);
    },


    /**
     * 分享QQ空间
     * 用法:
     * native.custom.shareQQZone({
        title: 'MAMA+|大V店',
        desc: 'MAMA+|大V店',
        link: 'MAMA+|大V店',
        imgUrl: 'http://pic.davdian.com/goods/1/20151017104524.png'
      })
     */
    shareQQZone: function shareQQZone() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '0',
        sharePlatform: '3'
      }, param);
      this.share(param);
    },


    /**
     * 分享到短信
     * 用法:
     * native.custom.shareSms({
        title: 'MAMA+|大V店',
        desc: 'MAMA+|大V店',
        link: 'MAMA+|大V店',
        imgUrl: 'http://pic.davdian.com/goods/1/20151017104524.png'
      })
     */
    shareSms: function shareSms() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '0',
        sharePlatform: '4'
      }, param);
      this.share(param);
    },


    /**
     * 复制分享链接
     * 用法:
     * native.custom.shareCopyUrl({
        link: 'MAMA+|大V店'
      })
     */
    shareCopyUrl: function shareCopyUrl() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '0',
        sharePlatform: '5'
      }, param);
      this.share(param);
    },


    /**
     * 分享图片
     * 用法:
     * native.custom.shareImg({
        show: '1',  // 1弹窗,0不弹窗
        shareType: '1',
        bigImageUrl: 'http://pic.davdian.com/free/back_top_icon_0803.png'  // imgUrl无效,必须用bigImageUrl
      })
     */
    shareImg: function shareImg() {
      var param = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      // 默认参数
      param = _jquery2.default.extend({
        show: '1', // 1弹窗,0不弹窗
        shareType: '1',
        bigImageUrl: 'http://pic.davdian.com/free/back_top_icon_0803.png' // imgUrl无效,必须用bigImageUrl
      }, param);
      this.share(param);
    },


    /**
     * 功能: 唤起app首页,不执行cmd
     * 用法:
     * native.custom.invoke()
     */
    invoke: function invoke() {
      call({
        v: '0.0.0',
        host: '',
        action: '',
        param: {
          invoke: true
        }
      });
    },


    /**
     * 功能: 唤起app并打开页面
     * 用法:
     * native.custom.openUrlWithApp({
        url:"https://nemo.davdian.com"
     })
     */
    openUrlWithApp: function openUrlWithApp(param) {
      call({
        v: '2.5.0',
        host: 'Browser',
        action: 'open',
        param: param
      }, true);
    }
  }
};

/**
 * @module dvd-service-js-native
 * @author swg [源码地址](http://gitlab.rd.vyohui.com/FE-Service/dvd-service-js-native.git)
 */
exports.default = native;
